storage/trace.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"context"
+	"os"
+
+	internalTrace "cloud.google.com/go/internal/trace"
+	"cloud.google.com/go/storage/internal"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	otelcodes "go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	storageOtelTracingDevVar = "GO_STORAGE_DEV_OTEL_TRACING"
+	defaultTracerName        = "cloud.google.com/go/storage"
+	gcpClientRepo            = "googleapis/google-cloud-go"
+	gcpClientArtifact        = "cloud.google.com/go/storage"
+)
+
+// isOTelTracingDevEnabled checks the development flag until experimental feature is launched.
+// TODO: Remove development flag upon experimental launch.
+func isOTelTracingDevEnabled() bool {
+	return os.Getenv(storageOtelTracingDevVar) == "true"
+}
+
+func tracer() trace.Tracer {
+	return otel.Tracer(defaultTracerName, trace.WithInstrumentationVersion(internal.Version))
+}
+
+// startSpan creates a span and a context.Context containing the newly-created span.
+// If the context.Context provided in `ctx` contains a span then the newly-created
+// span will be a child of that span, otherwise it will be a root span.
+func startSpan(ctx context.Context, name string, opts ...trace.SpanStartOption) (context.Context, trace.Span) {
+	// TODO: Remove internalTrace upon experimental launch.
+	if !isOTelTracingDevEnabled() {
+		ctx = internalTrace.StartSpan(ctx, name)
+		return ctx, nil
+	}
+	opts = append(opts, getCommonTraceOptions()...)
+	ctx, span := tracer().Start(ctx, name, opts...)
```

Comment:
Just to be clear, we are leaving out `cloud.google.com/go/` from the trace name?

I'd suggest at least adding it back to the internalTrace implementation.

Reply:
I've added the package name back to the internalTrace implementation with `appendPackageName()`. I originally removed this to follow the span name pattern of $package.$service.$method (c++ example `storage::Client::WriteObject/CreateResumableUpload`) but am open to adding it back. 

Pro is to have the official package name; Con is longer span name and repeated info with added attribute. Let me know what you think!

Reply:
I think  keeping `cloud.google.com/go/storage` in the span name is preferable.